distogram/__init__.py

```diff
@@ -262,106 +227,103 @@ def count_at(h, value):
-        ratio = (value - bins[-1][0]) / (h.max - bins[-1][0])
-        s = ratio * (bins[-1][1]) / 2
-        s += bins[-1][1] / 2
-        for j in range(len(bins) - 1):
-            s += bins[j][1]
+    v0, f0 = h.bins[0]
+    vl, fl = h.bins[-1]
+    if value <= v0:  # left
+        ratio = (value - h.min) / (v0 - h.min)
+        result = ratio * v0 / 2
+    elif value >= vl:  # right
+        ratio = (value - vl) / (h.max - vl)
+        result = (1 + ratio) * fl / 2
+        result += sum((f for _, f in h.bins[:-1]))
     else:
-        i = -1
-        while value > bins[i+1][0] and i < len(bins) - 1:
-            i += 1
+        i = sum(((value > v) for v, _ in h.bins)) - 1
+        vi, fi = h.bins[i]
+        vj, fj = h.bins[i + 1]
 
-        mb = bins[i][1] + (bins[i+1][1] - bins[i][1]) \
-            / (bins[i+1][0] - bins[i][0]) * (value - bins[i][0])
-        s = (bins[i][1] + mb) / 2 * (value - bins[i][0]) \
-            / (bins[i+1][0] - bins[i][0])
-        for j in range(i):
-            s = s + bins[j][1]
+        mb = fi + (fj - fi) / (vj - vi) * (value - vi)
+        result = (fi + mb) / 2 * (value - vi) / (vj - vi)
+        result += sum((f for _, f in h.bins[:i]))
 
-        s = s + bins[i][1] / 2
-    return s
+        result = result + fi / 2
 
+    return result
 
-def count(h):
-    '''Counts the number of elements in the distribution.
+
+def count(h: Distogram) -> float:
+    """ Counts the number of elements in the distribution.
 
     Args:
         h: A Distogram object.
 
     Returns:
         The number of elements in the distribution.
-    '''
-    return reduce(lambda acc, i: acc + i[1], h.bins, 0)
+    """
+    return reduce(lambda acc, b: acc + b[1], h.bins, 0)
 
 
-def bounds(h):
-    '''Returns the min and max values of the distribution.
+def bounds(h: Distogram) -> Tuple[float, float]:
+    """ Returns the min and max values of the distribution.
 
     Args:
         h: A Distogram object.
 
     Returns:
         A tuple containing the minimum and maximum values of the distribution.
-    '''
-    return (h.min, h.max)
+    """
+    return h.min, h.max
 
 
-def mean(h):
-    '''Returns the mean of the distribution.
+def mean(h: Distogram) -> float:
+    """ Returns the mean of the distribution.
 
     Args:
         h: A Distogram object.
 
     Returns:
         An estimation of the mean of the values in the distribution.
-    '''
+    """
     p = [i[0] for i in h.bins]
     m = [i[1] for i in h.bins]
     return _moment(p, m, 0, 1)
 
 
-def variance(h):
-    '''Returns the variance of the distribution.
+def variance(h: Distogram) -> float:
+    """ Returns the variance of the distribution.
 
     Args:
         h: A Distogram object.
 
     Returns:
         An estimation of the variance of the values in the distribution.
-    '''
+    """
     p = [i[0] for i in h.bins]
     m = [i[1] for i in h.bins]
-    return _moment(p, m, mean(h), 2)
+    return _moment(p, m, _moment(p, m, 0, 1), 2)
```

Comment:
Calling `_moment` directly for the mean lets us avoid having to recreate `p` and `m`.

Reply:
Excellent point on the zipped iterator. I've made the cahnge.